glommio/src/executor.rs

```diff
@@ -559,6 +559,276 @@ impl Default for LocalExecutorBuilder {
+
+impl LocalExecutorPoolBuilder {
+    /// Generates the base configuration for spawning a pool of
+    /// [`LocalExecutor`]s, from which configuration methods can be chained.
+    /// The method's only argument sets the number of [`LocalExecutor`]s to
+    /// spawn.
+    pub fn new(nr_shards: usize) -> Self {
+        Self {
+            nr_shards,
+            // spin_before_park: None,
+            name: String::from("unnamed"),
+            io_memory: 10 << 20,
+            preempt_timer_duration: Duration::from_millis(100),
+            placement: Placement::Unbound,
+        }
+    }
+
+    // /// Please see documentation under
+    // /// [`LocalExecutorBuilder::spin_before_park`] for details.  The setting
+    // /// is applied to all executors in the pool.
+    // pub fn spin_before_park(mut self, spin: Duration) -> Self {
+    //     self.spin_before_park = Some(spin);
+    //     self
+    // }
+
+    /// Please see documentation under [`LocalExecutorBuilder::name`] for
+    /// details. The setting is applied to all executors in the pool.
+    pub fn name(mut self, name: &str) -> Self {
+        self.name = String::from(name);
+        self
+    }
+
+    /// Please see documentation under [`LocalExecutorBuilder::io_memory`] for
+    /// details.  The setting is applied to all executors in the pool.
+    pub fn io_memory(mut self, io_memory: usize) -> Self {
+        self.io_memory = io_memory;
+        self
+    }
+
+    /// Please see documentation under [`LocalExecutorBuilder::preempt_timer`]
+    /// for details.  The setting is applied to all executors in the pool.
+    pub fn preempt_timer(mut self, dur: Duration) -> Self {
+        self.preempt_timer_duration = dur;
+        self
+    }
+
+    /// This method sets the [`Placement`] policy by which [`LocalExecutor`]s
+    /// are bound to the machine's hardware topology (i.e. which CPUs to
+    /// use).  The default is [`Placement::Unbound`].
+    pub fn placement(mut self, p: Placement) -> Self {
+        self.placement = p;
+        self
+    }
+
+    /// A method that generates a [`CPUSet`] according to the provided
+    /// [`Placement`] policy; sequential calls may generate different sets
+    /// depending on the [`Placement`]
+    fn make_cpu_set(&self) -> CPUSet {
+        match self.placement {
+            Placement::Unbound => CPUSet(None),
+        }
+    }
+
+    /// Spawn a pool of [`LocalExecutor`]s in a new thread according to the
+    /// [`Placement`] policy, which is `Unbound` by default.
+    ///
+    /// This method is the pool equivalent of [`LocalExecutorBuilder::spawn`].
+    ///
+    /// It takes a closure `fut_gen` which will be executed on each new thread
+    /// to obtain the [`Future`] to be executed there.  The `fut_gen`
+    /// closure is passed two arguments: a `usize` and a [`CPUSet`].  The
+    /// `usize` provides a sequential id of the threads created by this
+    /// [`LocalExecutorPoolBuilder`] (i.e. numbered in the range `0..nr_shards`
```

Comment:
I think we should avoid making promises about what the ID will be.
Once we connect channels to this (a certainly good add), we may want to use an ID that makes using channels simpler.

I'd take inspiration on the `std::thread` IDs, where we promise that an executor id is unique, but don't say anything about what to expect from it.

Reply:
I'll tie this to the notifier / reactor id (`sys::ReactorGlobalState`) for now.  Let me know if you would rather have a totally unique id and I'll create a new global.

Reply:
... and i'll change the thread id visible to the user from `usize` to some opaque `ThreadId` type

Reply:
Take a look at how the `mesh_channels` work. They assign a peer_id based on the order in which the executors were connected to the mesh.

Long term, I would like to provide a mesh by default with this group of executors and the Ids of the mesh and executor should match.

I actually think it is okay if they are of different types, if that gives us flexibility in implementation, as long as they compare to the same thing.

Reply:
I was still a bit unsure what the approach here should be since `std::thread::ThreadId` are globally unique but the ids created by `mesh_channels` would only be unique for each builder.  I've revised this, for now to just use a `usize`.

Reply:
A usize is fine, we just shouldn't promise what they are.

Also a per-builder id is perfectly fine. My idea is that once you are done with this.
Let's keep in mind that the main goal of this builder is convenience: so my idea is that the natural next step is to attach a Mesh builder automatically, that can be used to send messages to all other executors that were created together.

The per-builder id in this case is an advantage, not a liability: it allows for that id to be used to refer to message destinations without any translation.

Reply:
One other thing I was wondering was whether the `usize` id should be passed to `fut_gen` at all.  If we don't promise anything about what the `usize` is then the user might not have much use for it in their `fut_gen` (they're just random numbers to the user).  And we already have e.g. `Local::id()` that the user can use to for identification.  We could also provide an analogous `Local::mesh_id()`.

Reply:
right.

It makes sense that we don't pass it. We can already grab the current executor ID as a thread local, and can do the same with the mesh id when we get there